<a href="https://colab.research.google.com/github/naghma28/PlayStoreAppReviewAnalysis/blob/main/PlayStoreAppReviewAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name - Play Store App Review Analysis**

**Porject Type-** EDA

**Contribution-** Individual

## **Introduction**

In the today mobile world, Android accomadates the major market share, with more than 80% of the devices using Google applications. The Google Play Store is one of the largest and most used app store in the world till date. The Google Play store app data contains underlying hidden information that can have a significant impact on the app development industry. 

Our project's purpose is to analyze trends and patterns in the Android app industry and derive actionable insights for developers to assist them launch a successful app in a competitive market. Each row (labelled app) in the collection has values, ratings, size, and many other attributes. Another dataset comprises consumer reviews and feedback regarding the app, which is crucial in determining the app's importance.

The challenging task is to sift through the large dataset and discover the fundamental truth behind app development and success.


## **Problem Statement**

**BUSINESS PROBLEM OVERVIEW**

Analysing the Play Store app is critical for the success of the app-driven market. In the current context, customer reviews play a significant role in comprehending the app's functionality and usage. A thorough and detailed analysis will help developers in developing an app while keeping current trends and patterns in mind.


In this project, we will be addressing the following significant issues:


Problem 1: What are the top 10 most competing catagories in playstore ?

Problem 2: Which genres are the most popular and least popular in playstore ?

Problem 3: What are the frequency distributions of Rating, Price, Last Updated, Sentiment Polarity and Sentiment Subjectivity column(Mean value range) ?

Problem 4: Does last update year affects the average rating ?

Probblem 5: What are the percentage of paid and free apps ?

Problem 6: Determine the outliers in the rating column ?

Problem 7: Is there any correlation between Sentiment polarity/Sentiment subjectivity with Installs/Rating/Reviews ?

Problem 8: What are the top 5 apps that have generated the highest revenue ?


## **Business Objective**

Discover key factors responsible for app engagement and success.

# **Let's Begin**

# ***1. Knowing the Data***

We will begin our analysis, by importing important libraries and loading the dataset in our notebook.

## **Import libraries and load datasets**

In [223]:
# importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns

In [224]:
# mount drive 
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [225]:
# read file
filepath = '/content/drive/MyDrive/ColabNotebooks/Self Project/2. Play Store Review/Play Store Data.csv'
df_data = pd.read_csv(filepath)

df_review = pd.read_csv('/content/drive/MyDrive/ColabNotebooks/Self Project/2. Play Store Review/User Reviews.csv')

## **A Quick Glance at the Dataset**

**DATASET FIRST VIEW**

In [226]:
df_data.head(3)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   

  Reviews  Size    Installs  Type Price Content Rating  \
0     159   19M     10,000+  Free     0       Everyone   
1     967   14M    500,000+  Free     0       Everyone   
2   87510  8.7M  5,000,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up  
2               Art & Design    August 1, 2018       1.2.4  4.0.3 and up

In [227]:
df_review.head(3)

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN

**DATASET INFORMATION**

In [228]:
df_data.describe(include="all")

App Category       Rating Reviews                Size    Installs  \
count    10841    10841  9367.000000   10841               10841       10841   
unique    9660       34          NaN    6002                 462          22   
top     ROBLOX   FAMILY          NaN       0  Varies with device  1,000,000+   
freq         9     1972          NaN     596                1695        1579   
mean       NaN      NaN     4.193338     NaN                 NaN         NaN   
std        NaN      NaN     0.537431     NaN                 NaN         NaN   
min        NaN      NaN     1.000000     NaN                 NaN         NaN   
25%        NaN      NaN     4.000000     NaN                 NaN         NaN   
50%        NaN      NaN     4.300000     NaN                 NaN         NaN   
75%        NaN      NaN     4.500000     NaN                 NaN         NaN   
max        NaN      NaN    19.000000     NaN                 NaN         NaN   

         Type  Price Content Rating Genres    Last Updated  \
count   10840  10841          10840  10841           10841   
unique      3     93              6    120            1378   
top      Free      0       Everyone  Tools  August 3, 2018   
freq    10039  10040           8714    842             326   
mean      NaN    NaN            NaN    NaN             NaN   
std       NaN    NaN            NaN    NaN             NaN   
min       NaN    NaN            NaN    NaN             NaN   
25%       NaN    NaN            NaN    NaN             NaN   
50%       NaN    NaN            NaN    NaN             NaN   
75%       NaN    NaN            NaN    NaN             NaN   
max       NaN    NaN            NaN    NaN             NaN   

               Current Ver Android Ver  
count                10833       10838  
unique                2832          33  
top     Varies with device  4.1 and up  
freq                  1459        2451  
mean                   NaN         NaN  
std                    NaN         NaN  
min                    NaN         NaN  
25%                    NaN         NaN  
50%                    NaN         NaN  
75%                    NaN         NaN  
max                    NaN         NaN

In [229]:
df_review.describe(include="all")

App Translated_Review Sentiment  Sentiment_Polarity  \
count                 64295             37427     37432        37432.000000   
unique                 1074             27994         3                 NaN   
top     Angry Birds Classic              Good  Positive                 NaN   
freq                    320               247     23998                 NaN   
mean                    NaN               NaN       NaN            0.182146   
std                     NaN               NaN       NaN            0.351301   
min                     NaN               NaN       NaN           -1.000000   
25%                     NaN               NaN       NaN            0.000000   
50%                     NaN               NaN       NaN            0.150000   
75%                     NaN               NaN       NaN            0.400000   
max                     NaN               NaN       NaN            1.000000   

        Sentiment_Subjectivity  
count             37432.000000  
unique                     NaN  
top                        NaN  
freq                       NaN  
mean                  0.492704  
std                   0.259949  
min                   0.000000  
25%                   0.357143  
50%                   0.514286  
75%                   0.650000  
max                   1.000000

In [230]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [231]:
df_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


**ROWS & COLUMNS COUNT**

In [232]:
df_data.shape , df_review.shape

((10841, 13), (64295, 5))

### **Inference**

The project consist of two different datasets :

* Play Store App: It has 13 columns and 10481 rows with a combination of numerical and category data. It has information visible while examining the app's primary features.

* User Reviews: It has 5 columns and 64295 rows that deal with the emotions of each app rated by the user.

* The dataset contains a mix of category and numerical columns with specific information about the ratings, size, types, and features of the Google Play Store app, as well as positive and negative user evaluations.

* The datasets contain a few duplicate entries and NaN values that must be treated carefully before proceeding with the study.

## **Data Preprocessing**

Now that we have a general understanding of the dataset, we can see that there are some duplicates and missing values that must be addressed.

Cleaning the data for further analysis entails the following steps:

* Remove duplicate values
* Remove NaN values from the data
* Replace all the missing values with 0
* Dropping columns that are unnecessary and are insignificant for our analysis
* Correcting the data format for smooth analysis

**A. DUPLICATE VALUES**

In [233]:
# determing duplicate values in our dataset
df_data.duplicated().sum()

483

In [234]:
df_review.duplicated().sum()

33616

**DROPPING DUPLICATE VALUES**

In [235]:
# dropping the duplicate values from the play store app dataset and checking the count
mydata = df_data.drop_duplicates()
mydata.duplicated().sum()

0

In [236]:
# dropping the duplicate values from the user review dataset and checking the count
myreview = df_review.drop_duplicates()
myreview.duplicated().sum()

0

**B. MISSING VALUES / NULL VALUES**

In [237]:
# missing values/null values count
mydata.isnull().sum()

App                  0
Category             0
Rating            1465
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

In [238]:
myreview.isnull().sum()

App                         0
Translated_Review         987
Sentiment                 982
Sentiment_Polarity        982
Sentiment_Subjectivity    982
dtype: int64

In [239]:
# defining the function to create summarized dataframe

def get_info(df):
  col_df = pd.DataFrame(index = df.columns, columns = ["datatype", "total_count", "na_count", "non_na_count", "%na_count", "unique_count"])
  col_df["datatype"] = [type(i) for i in list(df.columns)]
  col_df["total_count"] = [len(df[i]) for i in list(df.columns)]
  col_df["na_count"] = [df[i].isnull().sum() for i in list(df.columns)]
  col_df["non_na_count"] = [(~df[i].isnull()).sum() for i in list(df.columns)]
  col_df["%na_count"] = [((df[i].isnull().sum())/len(df[i]))*100 for i in list(df.columns)]
  col_df["unique_count"] = [len(df[i].unique()) for i in list(df.columns)]
  return col_df

In [240]:
get_info(mydata)

datatype  total_count  na_count  non_na_count  %na_count  \
App             <class 'str'>        10358         0         10358   0.000000   
Category        <class 'str'>        10358         0         10358   0.000000   
Rating          <class 'str'>        10358      1465          8893  14.143657   
Reviews         <class 'str'>        10358         0         10358   0.000000   
Size            <class 'str'>        10358         0         10358   0.000000   
Installs        <class 'str'>        10358         0         10358   0.000000   
Type            <class 'str'>        10358         1         10357   0.009654   
Price           <class 'str'>        10358         0         10358   0.000000   
Content Rating  <class 'str'>        10358         1         10357   0.009654   
Genres          <class 'str'>        10358         0         10358   0.000000   
Last Updated    <class 'str'>        10358         0         10358   0.000000   
Current Ver     <class 'str'>        10358         8         10350   0.077235   
Android Ver     <class 'str'>        10358         3         10355   0.028963   

                unique_count  
App                     9660  
Category                  34  
Rating                    41  
Reviews                 6002  
Size                     462  
Installs                  22  
Type                       4  
Price                     93  
Content Rating             7  
Genres                   120  
Last Updated            1378  
Current Ver             2833  
Android Ver               34

In [241]:
get_info(myreview)

datatype  total_count  na_count  non_na_count  \
App                     <class 'str'>        30679         0         30679   
Translated_Review       <class 'str'>        30679       987         29692   
Sentiment               <class 'str'>        30679       982         29697   
Sentiment_Polarity      <class 'str'>        30679       982         29697   
Sentiment_Subjectivity  <class 'str'>        30679       982         29697   

                        %na_count  unique_count  
App                      0.000000          1074  
Translated_Review        3.217184         27995  
Sentiment                3.200887             4  
Sentiment_Polarity       3.200887          5411  
Sentiment_Subjectivity   3.200887          4475

**DEALING WITH MISSING VALUES / NULL VALUES**

Interestingly, we can now view extensive information about null values that exist across different parameters.

Based on the observations given, we can draw the following conclusions: 

* **PLAY STORE DATA** we have following null values:
      * Rating : 1465
      * Type & Content : 1
      * Android Ver : 8 
      * Current Ver : 3 

The rating column has many NaN values that can be replaced with the most prevalent value **MODE**.

We can drop these columns because the percentage of null values recorded in other columns is not statistically significant.

* **USER REVIEW DATA** we have following null values:
      * Translated Review : 982
      * Sentiment : 987
      * Sentiment Polarity : 987 
      * Sentiment Subjectivity : 987

 We can drop the columns of user review data with null values.

In [242]:
# sorting play store data in descending order wrt installs

mydata = mydata.sort_values(by="Installs", ascending = False)
mydata.head(3)

App            Category  Rating  \
10472  Life Made WI-Fi Touchscreen Photo Frame                 1.9    19.0   
4115                                   Twitter  NEWS_AND_MAGAZINES     4.3   
347              imo free video calls and chat       COMMUNICATION     4.3   

        Reviews                Size      Installs  Type     Price  \
10472      3.0M              1,000+          Free     0  Everyone   
4115   11657972  Varies with device  500,000,000+  Free         0   
347     4785892                 11M  500,000,000+  Free         0   

      Content Rating             Genres   Last Updated         Current Ver  \
10472            NaN  February 11, 2018         1.0.19          4.0 and up   
4115      Mature 17+   News & Magazines  July 30, 2018  Varies with device   
347         Everyone      Communication   June 8, 2018    9.8.000000010501   

              Android Ver  
10472                 NaN  
4115   Varies with device  
347            4.0 and up

In [243]:
# Life Made WI-Fi Touchscreen Photo Frame is on the top but is not the result of the huge downloads, so we drop it
# There must be error in entry
# dropping the column with index 10472

mydata.drop([10472], inplace = True)

In [244]:
# filling NaN values of the rating column with Mode

mydata["Rating"].fillna(mydata["Rating"].mode()[0], inplace = True)


In [245]:
# dropping insignificant columns with null values

print(mydata.shape)
mydata.dropna(how = 'any' , subset=["Type","Android Ver","Current Ver"], inplace = True)
print(mydata.shape)

(10357, 13)
(10346, 13)


In [246]:
# dropping null values from the columns of user reviews data

print(myreview.shape)
myreview.loc[myreview["Translated_Review"].isnull() | myreview["Sentiment"].isnull() | myreview["Sentiment_Polarity"].isnull() | myreview["Sentiment_Subjectivity"].isnull()]
myreview.dropna(how='any', subset=["Translated_Review", "Sentiment", "Sentiment_Polarity", "Sentiment_Subjectivity"], inplace=True)
print(myreview.shape)

(30679, 5)
(29692, 5)


<ipython-input-246-b9bb847a60c9>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  myreview.dropna(how='any', subset=["Translated_Review", "Sentiment", "Sentiment_Polarity", "Sentiment_Subjectivity"], inplace=True)


In [247]:
get_info(myreview)

datatype  total_count  na_count  non_na_count  \
App                     <class 'str'>        29692         0         29692   
Translated_Review       <class 'str'>        29692         0         29692   
Sentiment               <class 'str'>        29692         0         29692   
Sentiment_Polarity      <class 'str'>        29692         0         29692   
Sentiment_Subjectivity  <class 'str'>        29692         0         29692   

                        %na_count  unique_count  
App                           0.0           865  
Translated_Review             0.0         27994  
Sentiment                     0.0             3  
Sentiment_Polarity            0.0          5410  
Sentiment_Subjectivity        0.0          4474

Ultimately, we have successfully removed all duplicate entries and handled missing NaN values from the dataset.

**C. CORRECTING THE DATA TYPE**

We have effectively removed NaN values and duplicates from our dataset, but now we've noticed the use of some special character symbols and alphabet initials in our datasets, such as **'$' in Price, 'M', 'K' in Size, and '+', ','** in Installs, which can cause problems when performing statistical analysis. So it's ideal if we get rid of these undesirable elements first.


In [248]:
# removing $ sign from the price column

mydata['Price'] = mydata['Price'].apply(lambda p: p.replace("$","")  if "$" in str(p) else p)

#  dropping Everyone label from price column

mydata.drop(labels=mydata[mydata['Price']=='Everyone'].index , inplace = True)

# converting datatype of price from series to float

mydata['Price'] = mydata['Price'].apply(lambda p: float(p))

In [249]:
# checking the unique values of price
mydata['Price'].unique()

array([  0.  ,   1.99,   2.49,   2.99,   3.99,   0.99,   4.99,   5.99,
         8.99,   2.  ,   1.  ,  15.99,   3.49,   1.49,  17.99,   4.49,
         9.99,  11.99,   4.77,  29.99,  19.99, 399.99,   4.85,   2.6 ,
        14.99,   1.75,   1.04,   1.76,   6.99,   4.6 ,  39.99,   7.99,
         3.95,  13.99,  24.99,   9.  ,   2.9 ,   1.26,  19.9 ,  14.  ,
        12.99,   6.49,   3.88,   4.59,   5.  ,   3.02,   1.2 ,  16.99,
         3.28,   2.95,   1.97,   5.49,  10.  , 389.99,  18.99,   3.9 ,
       299.99,   1.59,   2.5 , 400.  ,   2.59,  46.99,   4.8 ,   3.04,
        25.99,  89.99,   3.61,   1.96,   8.49,   1.29,   4.29,   1.7 ,
         1.61,  15.46,  19.4 ,   4.84,  37.99, 379.99,   2.56,   1.5 ,
        33.99,  74.99,   7.49,   3.08,  10.99,  79.99, 109.99,  28.99,
        30.99, 394.99, 154.99, 200.  ])

In [250]:
# removing '+' and ',' from Installs

mydata['Installs'] = mydata['Installs'].apply(lambda x: x.replace("+","").replace(",","").strip() if "+" in str(x) or "," in str(x) else x)

# converting into integer

mydata['Installs'] = mydata['Installs'].apply(lambda x: int(x))

In [251]:
mydata['Installs'].unique()

array([ 500000000,     500000,        500,   50000000,      50000,
               50,    5000000,       5000,          5,  100000000,
           100000,        100,   10000000,      10000,         10,
       1000000000,    1000000,       1000,          1,          0])

In [254]:
# removing 'k' and 'M' from the Size columns
# converting kb into mb in the Size columns
# 1 mb = 1024 kb

mydata['Size'] =mydata['Size'].apply(lambda x: float(x.replace("k","").replace("M",""))/1024 if "k" in str(x) else x)
mydata['Size'] =mydata['Size'].apply(lambda x: float(x.replace("k","").replace("M","")) if "M" in str(x) else x)

# removing row of 'Varies with device' in size Column

mydata.drop(labels=mydata[mydata['Size'] == 'Varies with device'].index, inplace = True)

# converting to float
mydata['Size'] =mydata['Size'].apply(lambda x: float(x))

In [255]:
# typecasting the str type to timestamp in "Latest Updated" column

mydata["Last Updated"]= pd.to_datetime(mydata["Last Updated"])

Finally, we have completed some important data formatting, and our data appears to be decent and structured enough to run our analysis. Let's take a quick look at all of the modifications we made:

* We began by eliminating the '$' sign from the Price column and converting it to float data type.

* We also eliminated the '+' and ',' symbols from our Installs columns because they were unnecessary, and the column was transformed to integer type.

* We also did some unusual formatting with the Size column, first deleting the 'k' and 'M' initials and then changing the kb into mb units to bring the data into line.


Let us now thoroughly study the variable we will be dealing with.

# ***2. Understanding the Variable***

**VARIABLE DESCRIPTION**

**1. Play Store Review Data (Main Dataset)** : The file contains 13 columns and 10841 rows in total.

* App Category: Category of the app. This could be beauty, business, entertainment, education...etc.

* Rating: How users rate the app out of 5, with 1 being the lowest rating and 5 being the highest.

* Reviews: The number of user reviews each app has received.

* Size: The memory size needed to install the application.

* Installs: The number of times each application has been installed by users.

* Type: Whether the app is free or a paid app.

* Price: The price of the app.

* Content Rating: This column specifies the intended audience for the app. Can be for teens, mature audience, or everyone.

* Genres: The sub-category for each app. Example: for the Education category, this could be Education: Pretend Play, for example.

* Last Updated: Release date of the most recent update for the app.

* Current Ver: The app's current version.

* Android Ver: The oldest version of Android OS supported by the app.


**2. User Review Data:** This file contains the result of the sentiment analysis conducted by the dataset creator. It has 64,295 rows of data with the following columns:

* App : Name of the app.

* Translated_Review: Either the original review in English, or a translated version if the orignal review is in another language.

* Sentiment: The result of the sentiment analysis conducted on a review. The value is either Positive, Neutral, or Negative.

* Sentiment_Polarity: A value indicating the positivity or negativity of the sentiment, values range from -1 (most negative) to 1 (most positive).

* Sentiment_Subjectivity: A value from 0 to 1 indicating the subjectivity of the review. Lower values indicate the review is based on factual information, and higher values indicate the review is based on personal or public opinions or judgements.


# ***3. Data Wrangling***